hoodie-spark/src/test/java/HoodieJavaStreamingApp.java

```diff
@@ -0,0 +1,279 @@
+      cmd.usage();
+      System.exit(1);
+    }
+    cli.run();
+  }
+
+  /**
+   *
+   * @throws Exception
+   */
+  public void run() throws Exception {
+    // Spark session setup..
+    SparkSession spark = SparkSession.builder().appName("Hoodie Spark Streaming APP")
+        .config("spark.serializer",
+            "org.apache.spark.serializer.KryoSerializer").master("local[1]")
+        .getOrCreate();
+    JavaSparkContext jssc = new JavaSparkContext(spark.sparkContext());
+
+    // folder path clean up and creation, preparing the environment
+    FileSystem fs = FileSystem.get(jssc.hadoopConfiguration());
+    fs.delete(new Path(streamingSourcePath), true);
+    fs.delete(new Path(streamingCheckpointingPath), true);
+    fs.delete(new Path(tablePath), true);
+    fs.mkdirs(new Path(streamingSourcePath));
+
+    // Generator of some records to be loaded in.
+    HoodieTestDataGenerator dataGen = new HoodieTestDataGenerator();
+
+    List<String> records1 = DataSourceTestUtils.convertToStringList(
+        dataGen.generateInserts("001", 100));
+    Dataset<Row> inputDF1 = spark.read().json(jssc.parallelize(records1, 2));
+
+    List<String> records2 = DataSourceTestUtils.convertToStringList(
+        dataGen.generateUpdates("002", 100));
+
+    Dataset<Row> inputDF2 = spark.read().json(jssc.parallelize(records2, 2));
+
+    // setup the input for streaming
+    Dataset<Row> streamingInput = spark.readStream().schema(inputDF1.schema())
+        .json(streamingSourcePath);
+
+
+    // start streaming and showing
+    ExecutorService executor = Executors.newFixedThreadPool(2);
+
+    // thread for spark strucutured streaming
+    Future<Void> streamFuture = executor.submit(new Callable<Void>() {
+      public Void call() throws Exception {
+        logger.info("===== Streaming Starting =====");
+        stream(streamingInput);
+        logger.info("===== Streaming Ends =====");
+        return null;
+      }
+    });
+
+    // thread for adding data to the streaming source and showing results over time
+    Future<Void> showFuture = executor.submit(new Callable<Void>() {
+      public Void call() throws Exception {
+        logger.info("===== Showing Starting =====");
+        show(spark, fs, inputDF1, inputDF2);
+        logger.info("===== Showing Ends =====");
+        return null;
+      }
+    });
+
+    // let the threads run
+    streamFuture.get();
+    showFuture.get();
+
+    executor.shutdown();
+  }
+
+  /**
+   * Adding data to the streaming source and showing results over time
+   * @param spark
+   * @param fs
+   * @param inputDF1
+   * @param inputDF2
+   * @throws Exception
+   */
+  public void show(SparkSession spark,
+                   FileSystem fs,
+                   Dataset<Row> inputDF1,
+                   Dataset<Row> inputDF2) throws Exception {
+    inputDF1.write().mode(SaveMode.Append).json(streamingSourcePath);
+    // wait for spark streaming to process one microbatch
+    Thread.sleep(3000);
```

Comment:
Instead of arbitrary sleep times which can cause test flakiness, can you use test-function wait-Until-A-Condition ( https://github.com/awaitility/awaitility#Example_1_-_Simple)  wherever you use Thread.sleep